### PR TITLE
refactor(modules/vpc)!: VPC module refactor

### LINKED ISSUE #226: Refactor: VPC Module
### Is your feature request related to a problem?

- VPC module requires refactoring to properly handle multi-region/subnet deployments.

### Describe the solution you'd like

- Singleton type of approach for being able to call the module once and properly control everything from tfvars file.

### Describe alternatives you've considered.

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by horiagunica**
## Description

- Full refactor of VPC module to achieve singleton design pattern to simplify multi-region deployments
- Structure of new VPC module is the following :
  - VPC Network (created or read)
    - List of Subnets (created or read)
      - Per-region deployment of subnets
    - List of Firewall rules (created only)
- Added additional options to module flexibility :
  - Possibility of configuring dedicated firewall rules
  - Priority of each firewall rule
  - Source + Target tags firewall rules
  - Source + Target service accounts firewall rules
  - All firewall rules still maintain ingress only direction
- All examples have been refactored to use new module:
  - Updated `standalone_vmseries_with_metadata_bootstrap` example to use updated `for_each` iteration design
  - Heavily simplified `vpc_peering_common_with_network_tags` example due to new module region flexibility (code length was cut in half)
- Bumped minimum TF version to 1.3 for compatibility with variables object types.
- Added a new `multi_nic_common` example with full bootstrapping.
- This PR closes #226 

## Motivation and Context

- Current VPC module has design issues especially with multi-region designs (module needs to be called per region).

## How Has This Been Tested?

- Terratest `Idempotency` run for all examples.
- Traffic also tested for 3 examples.

## Types of changes

- Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes if appropriate.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by pavelrn** [comment]
I've just used the module in one the deployment. I've tested:
- VPC and subnets created by the module
- VPC exists, subnets are created by the module
- Using existing VPC in the module to implement VPC peering
all scenarios work well. 
Well done @horiagunica !

**Review by sebastianczech — APPROVED** [review]
Great job.
I have one comment regarding order of variables and outputs , maybe we can put them in order as we have order in tfvars e.g.:

```
    vpc_name                        = "fw-mgmt-vpc"
    create_network                  = true
    delete_default_routes_on_create = false
    mtu                             = "1460"
    routing_mode                    = "REGIONAL"
    subnetworks = {
      fw-mgmt-sub = ...
    }
    firewall_rules = {
      allow-mgmt-ingress = ...
    }
```
?

File: examples/panorama_standalone/variables.tf [hunk 1]
```diff
@@ -23,16 +23,31 @@ variable "networks" {
     Example of variable deployment :
 
     ```
-    vpcs = {
+    networks = {
```
Comment (pimielowski) [inline examples/panorama_standalone/variables.tf:26]: In AWS and Azurerm we named variable related to the virtual network on its service name, ie: in AzureRM it is vnets,  in AWS it is vpcs. I highly encourage you to change name of this variable to be familiar in all stacks :)

**Review by pimielowski — APPROVED** [review]
Good job! 👍

File: examples/panorama_standalone/variables.tf (hunk 1 above)
Comment (horiagunica) [reply examples/panorama_standalone/variables.tf:26]: I agree but there are a lot of names that point to it being network. Two of the most important:
- The terraform resource is called `google_compute_network`
- Within Google - you have `VPC networks`

So while in AWS it is indeed called `aws_vpc` - for google - they just stuck to `network` while VPC contains more components besides the network itself (VPC serverless access, shared VPC, etc.).

I would in google's case - the service name itself is `network` or `vpc_network` worst case. Does this reasoning seem ok with you ?

Comment (pimielowski) [reply examples/panorama_standalone/variables.tf:26]: It's fine for me than :)

File: examples/multi_nic_common/variables.tf [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+        }
+      }
+    }
+  }
+  ```
+
+  For a full list of available configuration items - please refer to [module documentation](https://github.com/PaloAltoNetworks/terraform-google-vmseries-modules/tree/main/modules/vpc#input_networks)
+
+  Multiple keys can be added and will be deployed by the code.
+  EOF
```
Comment (pimielowski) [inline examples/multi_nic_common/variables.tf:116]: Missing `type` and `default` parameter in variable

File: examples/multi_nic_common/variables.tf [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+    bootstrap_options = {
+      type                = "dhcp-client"
+      mgmt-interface-swap = "enable"
+    }
+  }
+  ``` 
+
+  Bootstrap options can be moved between vmseries individual instance variable (`vmseries`) and this common vmserie variable (`vmseries_common`).
+  EOF
+}
```
Comment (pimielowski) [inline examples/multi_nic_common/variables.tf:198]: Missing `type` and `default` parameter in variable

**Review by pimielowski — APPROVED** [review]
I found few variables without `type` and `default` parameters

File: modules/vpc/variables.tf
Comment (michalbil) [inline modules/vpc/variables.tf:1]: > I have one comment regarding order of variables and outputs (files variables.tf and outputs.tf), maybe we can put them in order as we have order in tfvars e.g.:

I'd propose to focus on variables usage/importance here, ie. keep required/often changed/provider level ones on top, others below:
```
project_id
name
create_network
subnetworks
firewall_rules
delete_default_routes_on_create
mtu
routing_mode
```

File: modules/vpc/variables.tf [hunk 4]
```diff
@@ -1,85 +1,152 @@
-variable "networks" {
+variable "subnetworks" {
   description = <<-EOF
-  Map of networks, a minimal example:
+  A map containing subnetworks configuration. Subnets can belong to different regions.
+  List of available attributes of each subnetwork entry:
+  - `subnetwork_name` : Name of the subnetwork.
```
Comment (michalbil) [inline modules/vpc/variables.tf:5]:
```suggestion
  - `name` : Name of the subnetwork.
```
It's within the same scope as `name` for network used previously, so maybe we can shorter this attribute name?

File: modules/vpc/variables.tf [hunk 5]
```diff
@@ -1,85 +1,152 @@
...
-      allowed_sources  = ["10.0.0.0/8", "98.98.98.0/28"]
-      allowed_protocol = "UDP"
-      allowed_ports    = ["53", "123-125"]
-      log_metadata     = "EXCLUDE_ALL_METADATA"
-      routing_mode     = "GLOBAL"
+variable "create_network" {
+  description = <<-EOF
+  A flag to indicate the creation or import of a VPC network.
+  Setting this to `true` will create a new network managed by terraform.
+  Setting this to `false` will try to read the existing network with those name and region settings.
```
Comment (michalbil) [inline modules/vpc/variables.tf:40]:
```suggestion
  Setting this to `true` will create a new network managed by Terraform.
  Setting this to `false` will try to read the existing network identified by `name` and `project` variables.
```

File: modules/vpc/variables.tf [hunk 6]
```diff
@@ -1,85 +1,152 @@
...
-  default     = []
-  type        = list(string)
+  List of available attributes of each firewall rule entry:
+  - `name` : Name of the firewall rule.
+  - `source_ranges` : (Optional) A list of strings containing the source IP ranges to be allowed on the firewall rule.
+  - `source_tags` : (Optional) A list of strings containing the source network tags to be allowed on the firewall rule.
+  - `source_service_accounts` : (Optional) A list of strings containg the source servce accounts to be allowed on the firewall rule.
+  - `target_service_accounts` : (Optional) A list of strings containing the service accounts for which the firewall rule applies to.
+  - `target_tags` : (Optional) A list of strings containing the network tags for which the firewall rule applies to. 
+  - `allowed_protocol` : The protocol type to match in the firewall rule. Possible values are : tcp, udp, icmp, esp, ah, sctp, ipip, all
```
Comment (michalbil) [inline modules/vpc/variables.tf:95]:
```suggestion
  - `allowed_protocol` : The protocol type to match in the firewall rule. Possible values are: `tcp`, `udp`, `icmp`, `esp`, `ah`, `sctp`, `ipip`, `all`.
```

File: modules/vpc/variables.tf [hunk 7]
```diff
@@ -1,85 +1,152 @@
...
+  - `ports` : A list of strings containing TCP or UDP port numbers to match in the firewall rule. This type of setting can be configured if allowing only TCP and UDP as protocols.
```
Comment (michalbil) [inline modules/vpc/variables.tf:96]:
```suggestion
  - `ports` : A list of strings containing TCP or UDP port numbers to match in the firewall rule. This type of setting can only be configured if allowing TCP and UDP as protocols.
```

File: modules/vpc/variables.tf [hunk 8]
```diff
@@ -1,85 +1,152 @@
...
+  - `priority` : (Optional) A priority value for the firewall rule. The lower the number - the more preffered the rule is.
```
Comment (michalbil) [inline modules/vpc/variables.tf:97]:
```suggestion
  - `priority` : (Optional) A priority value for the firewall rule. The lower the number - the more preferred the rule is.
```
Or "[...] the higher the priority of the rule."

File: modules/vpc/variables.tf [hunk 9]
```diff
@@ -1,85 +1,152 @@
...
+  }
+  validation {
+    condition = length(var.firewall_rules) > 0 ? alltrue([
+      for rule in var.firewall_rules : (
+        (rule.target_tags != null && rule.target_service_accounts == null) ||
+        (rule.target_tags == null && rule.target_service_accounts != null) ||
+        (rule.target_tags == null && rule.target_service_accounts == null)
+      )
+    ]) : true
+    error_message = "Please select only target_tags or target_service_accounts or neighter (apply to all instances in the network)."
```
Comment (michalbil) [inline modules/vpc/variables.tf:145]:
```suggestion
    error_message = "Please select only target_tags or target_service_accounts or neither (applies to all instances in the network)."
```

File: modules/vpc/variables.tf [hunk 10]
```diff
@@ -1,85 +1,152 @@
...
-  It is allowed from Terraform 0.14 onwards.
+variable "mtu" {
+  description = <<-EOF
+  MTU value for VPC Network. Acceptable values are between 1300 and 8896.
   EOF
+  default     = 1460
+  type        = number
+  validation {
+    condition     = var.mtu >= 1300 && var.mtu <= 8896
+    error_message = "MTU Range must be between 1300 and 8896 !"
```
Comment (michalbil) [inline modules/vpc/variables.tf:64]:
```suggestion
    error_message = "MTU value must be between 1300 and 8896."
```

File: modules/vpc/variables.tf [hunk 11]
```diff
@@ -1,85 +1,152 @@
...
+      ip_cidr_range = "192.168.0.0/24"
+      region = "us-east1"
     }
   }
   ```
+  EOF
+  default     = {}
+  type = map(object({
+    subnetwork_name   = string
+    create_subnetwork = bool
```
Comment (michalbil) [inline modules/vpc/variables.tf:25]:
```suggestion
    create_subnetwork = optional(bool, true)
```
I think we can stick to the same default as for the network.

File: modules/vpc/outputs.tf [hunk 12]
```diff
@@ -1,25 +1,10 @@
-output "networks" {
-  value = { for _, v in var.networks : v.name
-    => try(data.google_compute_network.this[v.name], google_compute_network.this[v.name], null)
-  }
+output "network" {
+  description = "Created or read network attributes."
+  value       = try(data.google_compute_network.this[0], google_compute_network.this[0])
 }
```
Comment (michalbil) [inline modules/vpc/outputs.tf:4]:
```suggestion
}

```

File: modules/vpc/variables.tf
Comment (horiagunica) [reply modules/vpc/variables.tf:1]: Done!

**Review by migara — APPROVED** [review]
Great work @horiagunica 🚀

File: modules/vpc/variables.tf (hunk 9 above)
Comment (horiagunica) [reply modules/vpc/variables.tf:115]: Modified the error message descriptions.

File: modules/vpc/variables.tf (hunk 4 above)
Comment (horiagunica) [reply modules/vpc/variables.tf]: Modified the module and all of the examples to use `name` for the subnetworks map.

**Review by michalbil — APPROVED** [review]

**Pull request merged**